Fix FakeTensor issues with copy_ between devices

jansel opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* #106234
* #106176
* #105887
* __->__ #106172
* #106233

Used to fail with:
```
RuntimeError: Unhandled FakeTensor Device Propagation for aten.copy_.default, found two different devices cpu, cuda:0
```

jansel committed:
Fix FakeTensor issues with copy_ between devices

Used to fail with:
```
RuntimeError: Unhandled FakeTensor Device Propagation for aten.copy_.default, found two different devices cpu, cuda:0
```

[ghstack-poisoned]

eellison commented on torch/_subclasses/fake_tensor.py:
```diff
@@ -1124,8 +1124,12 @@ def merge_devices(t):
 
-        tree_map(merge_devices, args)
-        tree_map(merge_devices, kwargs)
+        if func is aten.copy_.default and args and isinstance(args[0], FakeTensor):
```
register it here instead ? https://github.com/pytorch/pytorch/blob/2319e5f64075cf318b1edda839eac0723b2a10be/torch/_subclasses/fake_tensor.py#L455

jansel committed:
Update on "Fix FakeTensor issues with copy_ between devices"


Used to fail with:
```
RuntimeError: Unhandled FakeTensor Device Propagation for aten.copy_.default, found two different devices cpu, cuda:0
```

[ghstack-poisoned]

jansel committed:
Update on "Fix FakeTensor issues with copy_ between devices"


Used to fail with:
```
RuntimeError: Unhandled FakeTensor Device Propagation for aten.copy_.default, found two different devices cpu, cuda:0
```

[ghstack-poisoned]

jansel committed:
Update on "Fix FakeTensor issues with copy_ between devices"


Used to fail with:
```
RuntimeError: Unhandled FakeTensor Device Propagation for aten.copy_.default, found two different devices cpu, cuda:0
```

[ghstack-poisoned]

jansel committed:
Update on "Fix FakeTensor issues with copy_ between devices"


Used to fail with:
```
RuntimeError: Unhandled FakeTensor Device Propagation for aten.copy_.default, found two different devices cpu, cuda:0
```

[ghstack-poisoned]

jansel commented:
@pytorchbot merge
